go/shared.go

```diff
@@ -1,3 +1,38 @@
 package app
 
+import (
+	"log"
+
+	"go.temporal.io/sdk/workflow"
+)
+
 const BreakfastTaskQueue = "BREAKFAST_TASK_QUEUE"
+
+type Compensations []any
+
+func (s *Compensations) AddCompensation(activity any) {
+	*s = append(*s, activity)
+}
+
+func (s Compensations) Compensate(ctx workflow.Context, in_parallel bool) {
+	if !in_parallel {
+		for i := len(s) - 1; i >= 0; i-- {
+			errCompensation := workflow.ExecuteActivity(ctx, s[i]).Get(ctx, nil)
+			if errCompensation != nil {
+				log.Println("Executing compensation failed", errCompensation)
+			}
+		}
+	} else {
+		var pendingFutures []workflow.Future
+		for i := len(s) - 1; i >= 0; i-- {
+			f := workflow.ExecuteActivity(ctx, s[i])
+			pendingFutures = append(pendingFutures, f)
+		}
+		for _, pending := range pendingFutures {
+			if err := pending.Get(ctx, nil); err != nil {
```

Comment:
You can use a "Selector" if you'd rather wait on first finished instead of first started to finish. Otherwise, if the second one fails immediately but the first takes 2 hours, you won't get an error log for two hours.

But I'm ok with it as is too just for demonstration purposes (seems in some other samples we wait for parallel compensations before logging failure too)

Reply:
oooh yeah, great. Done!